src/raz_meta.rs

```diff
@@ -112,4 +119,67 @@ impl FirstLast for usize {
 	fn last() -> Self { usize::max_value() }
 }
 
-// TODO: HashMap
\ No newline at end of file
+/// Metadata for names in a raz tree.
+///
+/// Hash is implemented by a no-op, since the data here
+/// can be found elsewhere in the tree. This is not intended
+/// to be used outside of a raz.
+#[derive(Clone,Eq,PartialEq,Debug)]
+pub struct Names(pub HashMap<Name,()>);
+impl Hash for Names {
+	/// does nothing
+	fn hash<H:Hasher>(&self, _state: &mut H) {}
+}
+
+#[derive(Clone,Debug)]
+pub enum NameIndex {
```

Comment:
The presence of `FarLeft` and `FarRight` in type `NameIndex` is further evidence of my earlier observation: The `OnlyEnds` index type works for _all_ types of meta data.

The current design (with `Index` as an associated type of the `RazMeta` trait) does not let you factor out this orthogonal logic and instead seems to force you to reimplement it for each choice of meta data type.